### DISCUSSION TITLE
Milestones

### DISCUSSION THREAD

**Discussion by savchenko** [taskchampion/taskchampion#160] [General]
TC currently stands at v0.3.0. This is an attempt to untangle current issues and create a path towards v1.0.0
This list is based on the "tick-tock" model, starting with v0.4.0 being "tock" to stabilize the current state.

### v0.4.0

- [ ] Documentation is cleaned-up, broken links are fixed. https://github.com/taskchampion/taskchampion/issues/149
- [ ] CI is working for code and docs. https://github.com/taskchampion/taskchampion/issues/140
- [ ] APIs are versioned. https://github.com/taskchampion/taskchampion/issues/151
- [x] Deprecation/compatibility policies are defined. https://github.com/taskchampion/taskchampion/issues/210
- [ ] Document environment variables #100
- [ ] Limit modifications allowed in the add command #105
- [ ] Use a cache in CI to avoid complete rebuilds #33

### v0.5.0

- [ ] Colorization of CLI output where appropriate https://github.com/taskchampion/taskchampion/issues/116 & https://github.com/taskchampion/taskchampion/issues/86
- [ ] Deletion mechanism is ironed-out (https://github.com/djmitche/taskchampion/issues/26)
- [ ] Confirmation prompts https://github.com/taskchampion/taskchampion/issues/179
- [ ] Task modification logic https://github.com/taskchampion/taskchampion/issues/209
- [ ] Basic `fzf` integration (happy to help with this one as long as someone handles `stdin` on the "Rust side") #104 
- [ ] `task config` dumps configuration (shall it eventually provide an interactive config?)
- [ ] Implement "due date" (https://github.com/djmitche/taskchampion/issues/82)
- [ ] Implement recurrence (https://github.com/djmitche/taskchampion/issues/81)
- [ ] Implement waiting #83
- [ ] Implement synthetic tags #80

### v0.6.0

- [ ] Switch to `anyhow` and `thiserror` #148
- [ ] Grand old integration test #128
- [ ] Only derive encryption key once #129
- [ ] Test cli + sync-server in a grand old integration test #128
- [ ] Parse CLI args as Tag earlier #111
- [ ] Accept `--help` in more places https://github.com/taskchampion/taskchampion/issues/178
- [ ] Add cache-control headers to server #99
- [ ] Allow taking all remaining arguments as description, with `--` #135 

### v0.7.0

- [ ] Allow shorthands for report configuration #139
- [ ] Allow taking all remaining arguments as description #135
- [ ] Automatically document built-in reports #137 
- [ ] Implement "undo" #92
- [ ] Basic dependencies implementation (https://github.com/djmitche/taskchampion/issues/84)
- [ ] Automatically document built-in reports #137 

### v0.8.0

- [ ] Handle long URLs and (maybe) highlight them with OSC-8 (https://github.com/djmitche/taskchampion/issues/95)
- [ ] Ask user to confirm destructive actions (https://github.com/djmitche/taskchampion/issues/91)
- [ ] Pass task content to `$EDITOR` (https://github.com/djmitche/taskchampion/issues/97)
- [ ] Advanced `fzf` integration (https://github.com/djmitche/taskchampion/issues/104 et al.)
- [ ] Implement UDAs #90

### v0.9.0

- [ ] SQLite backend (related to https://github.com/djmitche/taskchampion/issues/131)
- [ ] Allow modifying status with status:.. #134
- [ ] Basic implementation of "urgency" (https://github.com/djmitche/taskchampion/issues/87)
- [ ] Maybe allow "expose" of the task list (https://github.com/djmitche/taskchampion/issues/159)
- [ ] Publicize this effort #52
- [ ] Allow machine-readable export #173

### v1.0.0

- [ ] Bugfixes and extensive testing
- [ ] Feedback collection, crash reports, etc.
- [ ] Strip debug symbols from release binary https://github.com/taskchampion/taskchampion/issues/208

### Future

- [ ] Mobile applications
- [ ] New website
- [x] New ID https://github.com/taskchampion/taskchampion/issues/207
- [ ] Happy and productive users around the world

If there are no objections, I'll go over the Issues List with a finer-tooth comb tomorrow and sort everything into the respective milestones.

**Comment by djmitche** [comment]
Is [this](https://en.wikipedia.org/wiki/Tick%E2%80%93tock_model) the tick-tock model?

This looks good.  Note that per semver, 0.x -> 0.(x+1) is considered a "major" version and can break API compatibility.  It might help to have some general guidelines for users at each stage, too: "ready for testing, please report bugs", "ready for day-to-day use, but may still have breaking changes", etc.  For example, switching to an sqlite backend will entail a breaking change for users.

When this is settled, I think the summaries would be good as descriptions on each milestone.

**Comment by dbr** [comment]
Depends on where we'd expect people to start using/testing the application - but I'd suspect that moving the switch to SQLite to an ealier milestone might make sense (v0.5?) - if we do it soon, there's no expectation that there would be a migration step to go from LMDB->SQlite.

**Reply by djmitche** [reply]
True -- there's a balance of stabilizing too soon, which may prevent us from solving problems in the most natural way; and stabilizing too late, which may be after users want to start using the app.

**Comment by savchenko** [comment]
OK, this is done. All issues are either:
1. Assigned to a milestone.
2. Tagged with [TBD](https://github.com/taskchampion/taskchampion/labels/TBD)

Latter means that issue requires further clarification, there is an unanswered question or it is simply depends on something else.

**Comment by djmitche** [comment]
I've copied the milestone descriptions from above into https://github.com/taskchampion/taskchampion/milestones.  Those might need a bit of updating, but they're close :)
